data-analytics/dataflow-python-examples/google/data_lake_to_mart_cogroupbykey.py

```diff
@@ -0,0 +1,195 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+""" data_lake_to_mart.py demonstrates a Dataflow pipeline which reads a 
+large BigQuery Table, joins in another dataset, and writes its contents to a 
+BigQuery table.  
+
+ To run this script, set up your python environment by running these commands:
+
+ virtualenv env
+ . ./env/bin/activate
+ pip install apache-beam
+ pip install apache_beam[gcp]
+ python ./google/data_ingestion.py --project=insert-your-project-here
+
+
+ Apache Beam setup instructions in more detail:
+    https://beam.apache.org/get-started/quickstart-py/
+
+ Virtual Env and Pip Command Explanation:
+  https://www.dabapps.com/blog/introduction-to-pip-and-virtualenv-python/
+"""
+
+from __future__ import absolute_import
+import argparse
+import csv
+import logging
+import os
+import apache_beam as beam
+from apache_beam.io.gcp.bigquery import parse_table_schema_from_json
+from apache_beam.options.pipeline_options import PipelineOptions
+from apache_beam.pvalue import AsDict
+
+
+class DataIngestion:
+    def __init__(self):
+        """"""
+        dir_path = os.path.dirname(os.path.realpath(__file__))
+        self.schema_str = ''
+        #this is the schem of output file
+        with open(dir_path + '/resources/orders_denormalized.json') \
+                as f:
+            data = f.read()
+            # Wrapping the schema in fields: is required for the BigQuery API
+            self.schema_str = '{"fields": ' + data + '}'
+
+
+
+
+def run(argv=None):
+    """The main function which creates the pipeline and runs it"""
+    # This object helps us parse command line arguments
+    parser = argparse.ArgumentParser()
+    # Here we add some specific command line arguments we expect.   S
+    # This defaults the output table in your BigQuery you'll have
+    # to create the example_data dataset yourself using bq mk temp
+    parser.add_argument('--output', dest='output', required=False,
+                        help='Output BQ table to write results to.',
+                        default='lake.orders_denormalized_cogroupbykey')
+
+    # Parse arguments from the command line
+    known_args, pipeline_args = parser.parse_known_args(argv)
+
+    # DataIngestion is a class we built in this script to hold the logic for
+    # transforming the file into a BigQuery table
+    data_ingestion = DataIngestion()
+
+    schema = parse_table_schema_from_json(data_ingestion.schema_str)
+    pipeline = beam.Pipeline(options=PipelineOptions(pipeline_args))
+
+    # This function performs the "join" of the two datasets.
+    def add_account_details((acct_number, data)):
+        import traceback
+        if len(data['account_details']) is 0:
```

Comment:
From Python style guide: For sequences (strings, lists, tuples), use the fact that empty sequences are false, so if seq: and if not seq: are preferable to if len(seq): and if not len(seq): respectively.